nservicebus/hosting/extensions-hosting.md

```diff
@@ -17,6 +17,7 @@ snippet: extensions-host-configuration
 
 This code will register the endpoint with the hosting infrastructure and automatically start and stop it based on the host's application lifetime.
 
+WARNING: `UseNServiceBus` must be specified before any other service (e.g. `ConfigureWebHostDefaults`) which requires access to the `IMessageSession`.
```

Comment:
I think that is OK given the MS DI based eco system reintroduces ordering problems everywhere (sorry for my slight rant)

Reply:
> I think that is OK given the MS DI based eco system reintroduces ordering problems everywhere (sorry for my slight rant)

yeah I'll raise an issue anyway to discuss this (I think we need to avoid the Lazy to not break the whole application at least).

Reply:
That's an interesting thought because the solution would essentially be that core needs to decorate the session and start the endpoint on demand if needed and do that in way the allows reentrancy without starting multiple times. But honestly I'm really on the fence whether it should. The whole service collection ecosystem is busted when it comes to ordering, see

> Make your Autofac registrations. Order is important!
https://autofaccn.readthedocs.io/en/latest/integration/netcore.html

the options pattern

> You can add multiple configuration providers. Configuration providers are available from NuGet packages and are applied in the order that they're registered. 

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-3.0

> An inconsistent options state may exist due to the ordering of service registrations.

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-3.0#accessing-options-during-startup

>   // Order added affects options setup order

https://offering.solutions/blog/articles/2017/02/07/the-difference-between-addmvc-and-addmvccore/

> Each IStartupFilter can add one or more middlewares in the request pipeline. The filters are invoked in the order they were added to the service container.

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/startup?view=aspnetcore-3.0#extend-startup-with-startup-filters

> The order that middleware components are added in the Startup.Configure method defines the order in which the middleware components are invoked on requests and the reverse order for the response. The order is critical for security, performance, and functionality.

https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/index?view=aspnetcore-3.0#middleware-order

Reply:
So in short the IServiceCollection ecosystem assumes ordering